lib/Analysis/Membar.cpp

```diff
@@ -216,6 +217,13 @@ void MembarAnalysis::update(Operation *op, BlockInfo *blockInfo,
         }
       }
     }
+    // If this op is may be signalling other threads asynchronously, make sure
```

Comment:
@Mogball Can you check out line 176. Does this op fit there?

```

  if (isa<triton::gpu::AsyncWaitOp, triton::nvidia_gpu::TMAStoreWaitOp>(op) &&
      !isa<gpu::BarrierOp>(op->getNextNode())) {
    // If the current op is an async wait and the next op is not a barrier we
    // insert a barrier op and sync
    builder->setInsertionPointAfter(op);
    insertBarrier(op, builder);
    blockInfo->sync();
    return;
  }
```

Reply:
It's a little bit different because we do actually want to check for pending shared memory transactions before inserting the barrier. For example, `tmem_store` followed by a `arrive_barrier` doesn't require a barrier because `tcgen05.st.wait` will synchronize the threads beforehand.

Reply:
OK, so basically if no pending transactions, we do not need to add a barrier IIUC